t/SocketWithLogging.hpp

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <boost/asio/buffers_iterator.hpp>
+#include <boost/system/error_code.hpp>
+#include <iostream>
+#include <utility>
+
+namespace bredis {
+namespace test {
+
+template <typename NextLayer> class SocketWithLogging {
+    NextLayer stream_;
+
+  public:
+    template <typename... Args>
+    SocketWithLogging(Args &&... args) : stream_(std::forward<Args>(args)...) {}
+
+    void cancel() {
+        std::cout << "{cancel}" << std::endl;
+        stream_.cancel();
+    }
+
+    template <typename BufferSequence>
+    void dump(const char *prefix, const BufferSequence &buffers) {
+        using Iterator = boost::asio::buffers_iterator<BufferSequence, char>;
+
+        Iterator it = Iterator::begin(buffers);
+        Iterator end = Iterator::end(buffers);
+        auto size = std::distance(it, end);
+
+        std::cout << "{" << prefix << " " << size << " bytes}[";
+        for (; it != end; it++) {
+            std::cout << *it;
```

Comment:
Sorry, I should have been more clear. You need a mutex for the whole application! The `SocketWithLogging` wrapper will never be called concurrently (because Asio requires using an implicit or explicit strand for all asynchronous operations). I'm talking about the case where you have two different wrappers, running in different threads (Asio allows this) and they both are trying to write to `std::cout` at the same time. Its not an easy problem to solve.